File: Chap1_Arrays&Strings/1.1Determine if a string has all unique characters/IsUnique.js
/**
 * @param {string} str
 * time:O(n), n is input string length
 * space:O(128)=O(1)
 */
function isUnique(str) {
    // Assume it's a ASCII string
    // We first create a map to record the chars in this array
    let m = new Array(128).fill(0);
    for (let i = 0; i < str.length; i++) {
        m[str.charCodeAt(i)] += 1;
        if (m[str.charCodeAt(i)] > 1) {
            return false;
        }
    }
    return true;
}

/**
 * @param {string} str
 */
function isU(str) {
    // Use bits to check uniqueness
    // GOTCHA: Only valid for 32 bits because bit operator can only manipulate 32 bits
    let checker = 0;
    for (let i = 0; i < str.length; i++) {
        let val = str.charCodeAt(i) - 'a'.charCodeAt(0);
        if ((checker & (1 << val)) > 0) {
            return false;
        }
        checker |= (1 << val);
    }
    return true;
}

File: Chap1_Arrays&Strings/1.1_IsUnique/IsUnique.java
class Main {
    public static void main(String[] args) {
        // test1
        System.out.println(isU("aaa") == false);
        System.out.println(isU("abc") == true);
        System.out.println(isU("abcA") == true);
        System.out.println(isU("abcAA") == false);
        // test2
        System.out.println(isUnique("aaa") == false);
        System.out.println(isUnique("abc") == true);
        System.out.println(isUnique("abcA") == true);
        System.out.println(isUnique("abcAA") == false);
    }

    // time:O(n), n is input string length
    // space:O(128)=O(1)
    // build a dictionary and check later
    static boolean isU(String str) {
        // Assume this is a ASCII string
        int[] m = new int[128];
        for (int i = 0; i < str.length(); i++) {
            m[str.charAt(i)]++;
            if (m[str.charAt(i)] > 1) {
                return false;
            }
        }
        return true;
    }

    static boolean isUnique(String str) {
        // Use bits to check uniqueness
        // GOTCHA: Only valid for 32 bits because bit operator can only manipulate 32
        // bits
        int checker = 0;
        for (int i = 0; i < str.length(); i++) {
            // get char code
            int val = str.charAt(i) - 'a';
            if ((checker & (1 << val)) > 0) {
                return false;
            }
            checker |= (1 << val);
        }
        return true;
    }
}
